### ISSUE TITLE
[FEATURE] Database version migration 

### ISSUE THREAD

**Issue by ComputerGenius152** [jdabtieu/CTFOJ#28]
## Is your feature request related to a problem? Please describe.
I'm always frustrated when the database breaks after an update and I have to remake it.

## Describe the solution you'd like
Ideally fewer and fewer breaking changes should be introduced as the project matures, but if possible a migration tool in helpers.py would be nice. 

## Describe alternatives you've considered
Either that, or better documentation about when breaking changes are introduced, or a more user-friendly error message. Should we look at semantic versioning?

## Additional context
N/A.

**Comment by jdabtieu** [comment]
Yep, I've been wanting to do this for a while now.
Semantic versioning would be nice. For each update that contains a database change, we can provide a script that automatically migrates the database format.

Ideally fewer and fewer breaking changes should be introduced, but the project is still in the early stages and I am still determining the best way to implement some key features.

I don't think we should put migration tools in helpers.py, but as mentioned above, a migration script in each new release.
It would ideally be something like
```
v1.2.3
===============
Warning: Database changes
If you are updating from v1.2.2, please run the following script. If you are updating from an older version, run all the database update scripts in chronological order from that version to this one.
script.py
` ``
do some stuff
` ``

actual release notes go here
```

**Comment by jdabtieu** [comment]
Alternatively, we could make an update script `update.py` which automatically makes the patches when run

**Comment by jdabtieu** [comment]
Also I should probably announce every time I make a db update, I haven't done that before

**Comment by BattleMage0231** [comment]
While we are on the topic of the database, I wanted to ask why the cs50 module is being used instead of other modules like SQLAlchemy that are more secure and integrate better with Flask? In the case of SQLAlchemy, there are extensions for Flask that handle database migrations as well.

**Comment by jdabtieu** [comment]
it was the easiest to use at the very beginning of the project, in a private repo. Of course, changing to SQLAlchemy would be nice now, and probably wouldn't be too hard to do, but there are more important changes to do at the moment.
If you'd like to start migrating in a separate branch though, that would be great

**Comment by ComputerGenius152** [comment]
- BattleMage0231 has indicated to me over Discord that they are working on it.

**Comment by BattleMage0231** [comment]
I've noticed one possible problem while doing some preliminary review on the code.

The database architecture makes it so that some columns and tables have to be dynamically created (i.e. problems_master, contest and contest info tables). This is hard to replicate in SQLAlchemy, as well as could lead to some dangerous behaviors in the future. In this case, there is a list of banned names for contest tables. I think that this is a slightly hacky solution and could result in migration problems as well. 

A possible solution is creating some sort of table like "contest_infos" that could have two fields that represent contest ID and problem ID within contest, respectively.

I believe changes should be made to the database architecture before migrating to SQLAlchemy. Also, you should consider moving the Python source files to an src folder to avoid clutter.

**Comment by jdabtieu** [comment]
> A possible solution is creating some sort of table like "contest_infos" that could have two fields that represent contest ID and problem ID within contest, respectively.

Yep, we could do that, and it would reduce the number of tables as well.

In terms of the banned names, I do have a local copy of a more thorough blacklist, that will (probably) be future proof for a long time. I'm going to implement that after sorting out my other assigned issues first.

In terms of problems_master and contestinfo tables, me and Anthony had a discussion about this a few weeks (months?) ago. We decided at that time that this was the easiest and best way to implement it. Obviously, if it could be a problem in the future, we should do as much as possible to future proof it, even if it means changing the structure of the table completely.

**Comment by BattleMage0231** [comment]
Thank you for clarifying. I have some proposed changes to the database that should maintain its performance, and make it more secure/easier to migrate.

### Contest Tables:
The contests table can stay the same. Instead of :cid and :cidinfo tables, there could be three tables contest_users, contest_solved, and contest_problems that are created by the following schema.
```SQL
CREATE TABLE 'contest_users' ('contest_id' varchar(32) NOT NULL, 'user_id' integer NOT NULL, 'points' integer NOT NULL DEFAULT (0) , 'lastAC' datetime); /* stores contest user information */
CREATE TABLE 'contest_solved' ('contest_id' varchar(32) NOT NULL, 'user_id' integer NOT NULL, 'problem_id' varchar(64) NOT NULL); /* stores contest solved problem information */
CREATE TABLE 'contest_problems' ('contest_id' varchar(32) NOT NULL, 'problem_id' varchar(64) NOT NULL, 'name' varchar(256) NOT NULL, 'point_value' integer NOT NULL DEFAULT (0), 'category' varchar(64), 'flag' varchar(256) NOT NULL); /* stores contest problem information */
```

All queries that are currently being made from the application should be able to be replicated using the new tables.

### Solved Problem Tables:
A similar restructuring could be applied to the problems_master table.
```SQL
CREATE TABLE 'problem_solved' ('user_id' integer NOT NULL, 'problem_id' varchar(64) NOT NULL); /* each row stores a problem that a user has solved */
```

Of course, this would take a while to implement as many of the queries currently used will have to be modified. But I find that having a growth mindset about changes like these as opposed to a fixed one is really helpful when working on larger projects.

**Comment by jdabtieu** [comment]
That looks good. It should also fix #38 and #39 as well.
@ComputerGenius152 thoughts?
I'm approving these changes.

**Comment by ComputerGenius152** [comment]
Everything looks good to me, after this goes through I can start to work on #4 as well.

**Comment by jdabtieu** [comment]
I'll look at writing the migration script after the pull request gets merged, but there is a nonzero chance that migration will be too difficult/sketchy and the recommendation will be to rebuild the database, since this is such a big change.

**Comment by jdabtieu** [comment]
Closed in #41
